mpd: 0.21.25 -> 0.22

tobim opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change
New version: https://raw.githubusercontent.com/MusicPlayerDaemon/MPD/v0.22/NEWS

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Closure size grows from 483.1 to 483.2 Mb.

tobim committed:
mpd: 0.21.25 -> 0.22

doronbehar commented on pkgs/servers/mpd/default.nix:
```diff
@@ -128,16 +136,15 @@ let
-        doxygen
-      ];
+      ]
+        ++ concatAttrVals features_ nativeFeatureDependencies;
```
If we enable docs by default, and we don't have [structuredAttrs](https://github.com/NixOS/nixpkgs/pull/72074) yet, I think we should also put them unconditionally in the nativeBuildInputs, and remove the whole logic behind it.

doronbehar replied:
The reason is https://github.com/NixOS/nixpkgs/pull/87940#issuecomment-629645759

tobim replied:
I created the `featureDependencies` attrset and the related functions in order to make the derivation easier to maintain. Structuring it this way obviates how dependencies are related to features, and makes it easier to spot mistakes. It was only turned into a customization mechanism because of a review request.
I'd be interested if anyone actually uses `mpdWithFeatures`, or `mpd-small` for that matter.

doronbehar replied:
I personally might have used mpdWithFeatures, if I had the time to setup my dream Nix based sound system :). Currently ARM support (for my Raspberry Pi) is still unsatisfying in Nixpkgs, unfortunately, so I don't use it (I'm using Debian (:sob:) on [BalenaOS](https://balena.io/)).

I'm personally "not afraid" of advanced Nix functions and alike when derivations are composed, I'm only trying to echo review comments I had in the past. Hence I won't make this issue a blocker for the update. Thanks for doing it anyway :).

doronbehar commented on pkgs/servers/mpd/default.nix:
```diff
@@ -30,13 +30,17 @@
     # Input plugins
     curl          = [ curl ];
+    # Disabled because the mpd binary has the wrong output of liburing in its
+    # RUNPATH
```
Could you elaborate please?

tobim replied:
If `io_uring` is enabled, the build itself succeeds, but the tests fail because the binary isn't built correctly. `liburing.so` isn't found because the wrong output of `liburing` is inserted into the `RUNPATH` of the `mpd` binary. The meson wrapper of `pkg-config` seems to be unable to deal with the way `liburing` is installed in nixpkgs. If I remove the `lib` output of `liburing`, then everything builds, but I don't want to do that in this PR because it would trigger a much larger rebuild.

doronbehar replied:
> f io_uring is enabled, the build itself succeeds, but the tests fail because the binary isn't built correctly. liburing.so isn't found because the wrong output of liburing is inserted into the RUNPATH of the mpd binary. The meson wrapper of pkg-config seems to be unable to deal with the way liburing is installed in nixpkgs. If I remove the lib output of liburing, then everything builds.

That's because `liburing` is not built properly (cc @thoughtpolice):

```
$ tree $(nix-build --no-out-link -A liburing.out)
/nix/store/vvjjmlhnfph4ah8zv2dzyc6i7ccp6404-liburing-0.7
├── bin
│   ├── io_uring-cp
│   ├── io_uring-link-cp
│   ├── io_uring-test
│   └── io_uring-ucontext-cp
└── lib
    ├── liburing.a
    └── liburing.so -> /nix/store/pi16ynlm7fyvxxsq7xrd3s2k40ryn9ss-liburing-0.7-lib/lib/liburing.so.1.0.7

2 directories, 6 files
```

`liburing.so` is referenced between the "lib" and "out" outputs, that's why disabling "lib" worked for you. This is an issue with liburing and not mpd. 

> because it would trigger a much larger rebuild.

Fortunately, `liburing` is used only by mpd and samba, which luckily both have NixOS tests we can use :) 

I'm pushing a change to liburing, for which I tested samba to at least build with it, and I checked that liburing support was included:

```
$ nix-store --read-log $(nix-build --no-out-link -A samba) | grep uring
configuring
Checking for liburing package                                                                   : yes 
Checking for header liburing.h                                                                  : yes 
Checking for library uring                                                                      : yes 
Checking for io_uring_ring_dontfork in uring                                                    : ok 
VFS_SHARED: vfs_recycle,vfs_audit,vfs_extd_audit,vfs_full_audit,vfs_fake_perms,vfs_default_quota,vfs_readonly,vfs_cap,vfs_expand_msdfs,vfs_shadow_copy,vfs_shadow_copy2,vfs_readahead,vfs_xattr_tdb,vfs_streams_xattr,vfs_streams_depot,vfs_acl_xattr,vfs_acl_tdb,vfs_preopen,vfs_catia,vfs_media_harmony,vfs_unityed_media,vfs_fruit,vfs_shell_snap,vfs_commit,vfs_worm,vfs_crossrename,vfs_linux_xfs_sgid,vfs_time_audit,vfs_offline,vfs_virusfilter,vfs_syncops,vfs_dirsort,vfs_fileid,vfs_aio_fork,vfs_io_uring,vfs_aio_pthread,vfs_gpfs,vfs_btrfs,vfs_glusterfs_fuse
[3167/3704] Compiling source3/modules/vfs_io_uring.c
[3179/3704] Linking bin/default/source3/modules/libvfs_module_io_uring.so
[3765/3818] Compiling docs-xml/manpages/vfs_io_uring.8.xml
docs-xml/manpages/vfs_io_uring.8.xml:2: warning: failed to load external entity "http://www.oasis-open.org/docbook/xml/4.2/docbookx.dtd"
Note: Writing vfs_io_uring.8
[4004/4735] Linking bin/default/source3/modules/libvfs_module_io_uring.inst.so
[4182/4735] Linking bin/default/source3/modules/libvfs_module_io_uring.so
+ install /nix/store/lzz0jrwabnxfj9j2dmwgnv5mgznhmr81-samba-4.12.5/lib/samba/vfs/io_uring.so (from bin/default/source3/modules/libvfs_module_io_uring.inst.so)
+ install /nix/store/0jnicddaxyrbg3mdicgpdygir8bxp5fb-samba-4.12.5-man/share/man/man8/vfs_io_uring.8 (from bin/default/docs-xml/manpages/vfs_io_uring.8)
shrinking /nix/store/lzz0jrwabnxfj9j2dmwgnv5mgznhmr81-samba-4.12.5/lib/samba/vfs/io_uring.so
```

doronbehar commented on pkgs/servers/mpd/default.nix:
```diff
@@ -30,13 +30,17 @@
 
 let
   lib = stdenv.lib;
+  concatAttrVals = nameList: set: lib.concatMap (x: set.${x} or []) nameList;
```
If the nativeFeatureDependencies logic will be removed, I would find this wrapper function redundant if used only once :).

doronbehar committed:
liburing: Fix references between outputs issue

Experienced in:
https://github.com/NixOS/nixpkgs/pull/98869#discussion_r495492213

Set configure flags properly, don't override configurePhase, per:
https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/explicit-phases.md

Use different output names - use $bin for examples, and $out for
libraries which expectedly should be referenced by reverse deps.

doronbehar commented:
MPD now builds with liburing support, so I expect ofborg's MPD NixOS tests to pass.

tobim commented:
@doronbehar I don't have anything to add here.

doronbehar commented:
Also @thoughtpolice doesn't express any objection, so I think this is good to go.

doronbehar commented:
[This push](https://github.com/NixOS/nixpkgs/pull/98869#event-3830267657) fixed a typo in commit message.

lopsided98 commented:
This broke cross compilation of liburing, see #101810 for a fix.
